pkg/cmd/roachtest/tests/db_console_endpoints.go

```diff
@@ -181,53 +181,54 @@ func testEndpoint(
 	var resp *http.Response
-	switch ep.Method {
-	case http.MethodGet:
-		resp, err = client.Get(ctx, fullURL)
-	default:
-		return errors.Newf("unsupported HTTP method: %s", ep.Method)
-	}
-
-	if err != nil {
-		return errors.Wrapf(err, "failed to %s %s", ep.Method, fullURL)
-	}
-	defer resp.Body.Close()
-
-	// Note(alyshan): In multiversion tests, some endpoints may be unavailable
-	// as we may be acting as a secondary tenant (401, 501), or the database is on a version
-	// that does not support the endpoint (409).
-	if multiVersionTest {
-		// Since nodes get restarted in multiversion tests there is a race where certain nodes have
-		// not registered certain routes yet. Retry a few times to handle transient 404s.
-		if resp.StatusCode == http.StatusNotFound {
-			r := retry.StartWithCtx(ctx, retry.Options{
-				MaxRetries: 10,
-			})
-			for r.Next() {
-				resp, err = client.Get(ctx, fullURL)
-				if err != nil {
-					return errors.Wrapf(err, "failed to %s %s", ep.Method, fullURL)
-				}
-				if resp.StatusCode != http.StatusNotFound {
-					break
-				}
+	f := func() error {
+		switch ep.Method {
+		case http.MethodGet:
+			resp, err = client.Get(ctx, fullURL)
+		default:
+			return errors.Newf("unsupported HTTP method: %s", ep.Method)
+		}
+		if err != nil {
+			return errors.Wrapf(err, "failed to %s %s", ep.Method, fullURL)
+		}
+		defer resp.Body.Close()
+
+		// Note(alyshan): In multiversion tests, some endpoints may be unavailable
+		// as we may be acting as a secondary tenant (401, 501), or the database is on a version
+		// that does not support the endpoint (409).
+		if multiVersionTest {
+			if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusNotImplemented ||
+				resp.StatusCode == http.StatusConflict {
+				l.Printf("%v returned %v in multiversion testing mode", fullURL, resp.StatusCode)
+				return nil
 			}
 		}
 
-		// TODO(alyshan): Consider explicitly noting which endpoints fall into this category.
-		if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusNotImplemented ||
-			resp.StatusCode == http.StatusConflict {
-			l.Printf("%v returned %v in multiversion testing mode", fullURL, resp.StatusCode)
-			return nil
+		err = verifyResponse(resp)
+		if err != nil {
+			return errors.Wrap(err, "failed to verify response")
 		}
-	}
 
-	err = verifyResponse(resp)
-	if err != nil {
-		return errors.Wrap(err, "failed to verify response")
+		return nil
 	}
 
-	return nil
+	return withRetries(ctx, retry.Options{MaxRetries: 10}, f)
+}
+
+// withRetries runs the given function f with the provided retry options.
+// If f returns nil, the retry loop breaks and nil is returned.
+// If f returns an error and retries are exhausted, the last error is returned.
+func withRetries(ctx context.Context, opts retry.Options, f func() error) error {
+	var lastErr error
+	r := retry.StartWithCtx(ctx, opts)
+	for r.Next() {
+		if err := f(); err != nil {
+			lastErr = err
```

Comment:
It might be worth doing something like `lastErr = errors.join(lastErr, err)` or something to keep track of all the errors. What do you think?